app.rb

```diff
@@ -16,15 +59,280 @@ class MyApp < Sinatra::Base
+      redirect '/password'
+    end
+  end
+
+# if login info is not found redirect to new user page
+  post '/' do
+    u = User.find_by(email: params[:username])
+    if u && u.password == Digest::SHA256.hexdigest(params[:password])
+      login_user u
+      redirect '/dashboard'
+    else
+      @failed_login = true
+      erb :login
+    end
+  end
+
+
+  post '/logout' do
+    logout
+    redirect '/'
+  end
+
+  # create new user info
+  #puts Digest::SHA256.hexdigest "Hello World"
+  post '/newuser' do
+    u = User.create!(email: params[:username], password: params[:password])
+    login_user u
+    session[:message] = "Account Creation Successful"
+    redirect '/dashboard'
+  end
+
+  # new user page show
+  get '/newuser' do
+    erb :newuser
+  end
+
+  # get songs info for dashboard
+  get '/dashboard' do
+    @songs = Playlist.current_alpha_hash
+    erb :dashboard
+  end
+
+  post '/invite' do
+    unless params[:email] == ""
+    Pony.mail :to => params[:email],
+              :from => "friend@alphatunez.herokuapp.com",
+              :headers => { 'Content-Type' => 'text/html' },
+              :subject => "Welcome to AlphatuneZ!",
+              :body => body = erb(:invite_email, layout: false )
+    end
+    session[:message] = "Email sent to #{params[:email]}"
+    redirect '/dashboard'
+  end
+
+  post "/user/song/vote" do
+    song = Song.find(params[:song_id])
+
+    if params[:vote] == "up"
+      vote_val = 1
     else
-      # raise e
-      puts e.message
+      vote_val = -1
     end
+
+    if v = Vote.find_by(user_id: current_user.id, song_id: song.id)
+      v.value = vote_val
+    else
+      v = Vote.create!(
+        user_id: current_user.id,
+        song_id: song.id,
+        value: vote_val,
+        placed_at: DateTime.now
+      )
+    end
+    redirect '/dashboard'
+  end
+
+  def login_user user
+    session[:logged_in_user_id] = user.id
+    session[:message] = "Login Successful"
+  end
+
+  def logout
+    session[:message] = "Logout Successful"
+    session.delete :logged_in_user_id
+    session.delete :token_handler
+    session.delete :state
+  end
+
+  def change_password newpassword
+    c = current_user
+    c.password = Digest::SHA256.hexdigest(newpassword)
+    c.save!
+  end
+
+  def token_string
+    session[:token_handler].access_token_type +
+      " " +
+      session[:token_handler].access_token
+  end
+
+  def current_user
+    if id = session[:logged_in_user_id]
+      User.find_by id: id
+    else
+      nil
+    end
+  end
+
+  def parsed_body
+    begin
+      @parsed_body ||= JSON.parse request.body.read
+    rescue
+      halt 400
+    end
+  end
+
+  get "/callback" do
+    if Digest::SHA256.hexdigest(session[:state]).to_s == params["state"]
+      session[:token_handler].request_refresh_and_access_tokens params
+      redirect "/dashboard"
+    else
+      halt 401
+    end
+  end
+
+  get "/api/me" do
+    session[:token_handler] = SpotifyApiToken.new
```

Comment:
Because of the underlying cookie structure, it makes me nervous to store raw Ruby objects here. I'd rather store the strings you need in the session and then generate a handler object when you need to use it (e.g. `SpotifyApiToken.new(session[:token]).request_and_access_tokens params`, where the `:token` itself is just a string)